File: examples/python/et0_calculator.py
import math


class ET0Calculator:
    
    def __init__(self, wind_speed=2.0, altitude=0.0):
        self.wind_speed = wind_speed
        self.altitude = altitude
        self.pressure = self._calculate_pressure()
        self.gamma = 0.000665 * self.pressure

    
    def _calculate_pressure(self):
        return 101.3 * pow((293.0 - 0.0065 * self.altitude) / 293.0, 5.26)

    
    def saturation_vapor_pressure(self, temperature):
        return 0.6108 * math.exp((17.27 * temperature) / (temperature + 237.3))

    
    def slope_vapor_pressure(self, temperature):
        es = self.saturation_vapor_pressure(temperature)
        return (4098 * es) / pow((temperature + 237.3), 2)

    
    def actual_vapor_pressure(self, temperature, relative_humidity):
        return (relative_humidity / 100.0) * self.saturation_vapor_pressure(temperature)

    
    def calculate(self, temperature, humidity, radiation):
        if not self._validate_inputs(temperature, humidity, radiation):
            return -1
        
        delta = self.slope_vapor_pressure(temperature)
        es = self.saturation_vapor_pressure(temperature)
        ea = self.actual_vapor_pressure(temperature, humidity)
        
        numerator = (0.408 * delta * radiation + 
                    self.gamma * (900 / (temperature + 273.15)) * 
                    self.wind_speed * (es - ea))
        
        denominator = delta + self.gamma * (1 + 0.34 * self.wind_speed)
        
        if denominator == 0:
            return -1
        
        et0 = numerator / denominator
        
        return et0 if 0 <= et0 <= 15 else -1

    
    def _validate_inputs(self, temp, humidity, radiation):
        if temp < -10 or temp > 50:
            return False
        if humidity < 0 or humidity > 100:
            return False
        if radiation < 0:
            return False
        return True



def main():
    calculator = ET0Calculator(wind_speed=1.5, altitude=15.0)
    
    test_cases = [
        {"temp": 25.0, "humidity": 60.0, "radiation": 8.0},
        {"temp": 30.0, "humidity": 50.0, "radiation": 10.0},
        {"temp": 35.0, "humidity": 40.0, "radiation": 12.0},
        {"temp": 20.0, "humidity": 70.0, "radiation": 6.0},
    ]
    
    print("ET0 Calculation Examples")
    print("-" * 50)
    
    for case in test_cases:
        et0 = calculator.calculate(
            case["temp"], 
            case["humidity"], 
            case["radiation"]
        )
        
        print(f"Temp: {case['temp']}C, Humidity: {case['humidity']}%, "
              f"Radiation: {case['radiation']} -> ET0: {et0:.2f} mm/day")



if __name__ == "__main__":
    main()

File: examples/c/et0_calculator.c
#include <stdio.h>
#include <math.h>
#include <stdbool.h>


typedef struct {
    double wind_speed;
    double altitude;
    double pressure;
    double gamma;
} ET0Calculator;


void et0_init(ET0Calculator *calc, double wind_speed, double altitude) {
    calc->wind_speed = wind_speed;
    calc->altitude = altitude;
    calc->pressure = 101.3 * pow((293.0 - 0.0065 * altitude) / 293.0, 5.26);
    calc->gamma = 0.000665 * calc->pressure;
}


double saturation_vapor_pressure(double temperature) {
    return 0.6108 * exp((17.27 * temperature) / (temperature + 237.3));
}


double slope_vapor_pressure(double temperature) {
    double es = saturation_vapor_pressure(temperature);
    return (4098 * es) / pow((temperature + 237.3), 2);
}


double actual_vapor_pressure(double temperature, double relative_humidity) {
    return (relative_humidity / 100.0) * saturation_vapor_pressure(temperature);
}


bool validate_inputs(double temp, double humidity, double radiation) {
    if (temp < -10 || temp > 50) return false;
    if (humidity < 0 || humidity > 100) return false;
    if (radiation < 0) return false;
    return true;
}


double calculate_et0(ET0Calculator *calc, double temperature, 
                     double humidity, double radiation) {
    if (!validate_inputs(temperature, humidity, radiation)) {
        return -1;
    }
    
    double delta = slope_vapor_pressure(temperature);
    double es = saturation_vapor_pressure(temperature);
    double ea = actual_vapor_pressure(temperature, humidity);
    
    double numerator = 0.408 * delta * radiation + 
                       calc->gamma * (900 / (temperature + 273.15)) * 
                       calc->wind_speed * (es - ea);
    
    double denominator = delta + calc->gamma * (1 + 0.34 * calc->wind_speed);
    
    if (denominator == 0) {
        return -1;
    }
    
    double et0 = numerator / denominator;
    
    return (et0 >= 0 && et0 <= 15) ? et0 : -1;
}


int main() {
    ET0Calculator calc;
    et0_init(&calc, 1.5, 15.0);
    
    typedef struct {
        double temp;
        double humidity;
        double radiation;
    } TestCase;
    
    TestCase test_cases[] = {
        {25.0, 60.0, 8.0},
        {30.0, 50.0, 10.0},
        {35.0, 40.0, 12.0},
        {20.0, 70.0, 6.0}
    };
    
    int num_cases = sizeof(test_cases) / sizeof(TestCase);
    
    printf("ET0 Calculation Examples\n");
    printf("--------------------------------------------------\n");
    
    for (int i = 0; i < num_cases; i++) {
        double et0 = calculate_et0(&calc, 
                                   test_cases[i].temp,
                                   test_cases[i].humidity,
                                   test_cases[i].radiation);
        
        printf("Temp: %.1fC, Humidity: %.1f%%, Radiation: %.1f -> ET0: %.2f mm/day\n",
               test_cases[i].temp,
               test_cases[i].humidity,
               test_cases[i].radiation,
               et0);
    }
    
    return 0;
}
